didman/api/v1/api.go

```diff
@@ -94,31 +94,50 @@ func (w *Wrapper) Routes(router core.EchoRouter) {
 	}))
 }
 
-// AddEndpoint handles calls to add a service. It only checks params and sets the correct return status code.
-// didman.AddEndpoint does the heavy lifting.
-func (w *Wrapper) AddEndpoint(ctx context.Context, request AddEndpointRequestObject) (AddEndpointResponseObject, error) {
-	id, err := did.ParseDID(request.Did)
+func (w *Wrapper) addOrUpdateEndpoint(
+	ctx context.Context, requestDID string, properties EndpointProperties,
+	operation func(ctx context.Context, id did.DID, serviceType string, endpoint url.URL) (*did.Service, error),
+) (*did.Service, error) {
+	id, err := did.ParseDID(requestDID)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(strings.TrimSpace(request.Body.Type)) == 0 {
+	if len(strings.TrimSpace(properties.Type)) == 0 {
 		return nil, core.InvalidInputError("invalid value for type")
 	}
 
-	u, err := url.Parse(request.Body.Endpoint)
+	endpoint, err := url.Parse(properties.Endpoint)
 	if err != nil {
 		return nil, core.InvalidInputError("invalid value for endpoint: %w", err)
 	}
+	return operation(ctx, *id, properties.Type, *endpoint)
+}
 
-	endpoint, err := w.Didman.AddEndpoint(ctx, *id, request.Body.Type, *u)
+// AddEndpoint handles calls to add a service. It only checks params and sets the correct return status code.
+// didman.AddEndpoint does the heavy lifting.
+func (w *Wrapper) AddEndpoint(ctx context.Context, request AddEndpointRequestObject) (AddEndpointResponseObject, error) {
+	endpoint, err := w.addOrUpdateEndpoint(ctx, request.Did, *request.Body, w.Didman.AddEndpoint)
 	if err != nil {
 		return nil, err
 	}
-
 	return AddEndpoint200JSONResponse(*endpoint), nil
 }
 
+// UpdateEndpoint handles calls to update a service. It only checks params and sets the correct return status code.
+// didman.UpdateEndpoint does the heavy lifting.
+func (w *Wrapper) UpdateEndpoint(ctx context.Context, request UpdateEndpointRequestObject) (UpdateEndpointResponseObject, error) {
+	if request.Body.Type != "" && request.Body.Type != request.Type {
```

Comment:
duplicate logic with addOrUpdateEndpoint checks?

Reply:
That check isn't there in `addOrUpdateEndpoint`. You mean with `UpdateCompoundService`? That's a different type, so it looks the same, but according to Go, it isn't